File: sdk/python/examples/berserker.py
"""Berserker Bot - All Attack, No Brain.

A simple aggressive bot that charges the closest enemy and attacks relentlessly.
Good starter example for learning the ArenaBot SDK.

Usage: python berserker.py <api_key> [server_url]
"""

import asyncio
import sys

from arena_sdk import ArenaBot
from arena_sdk.helpers import distance


class BerserkerBot(ArenaBot):
    """Pure aggression bot: move toward closest enemy, attack, never retreat."""

    def __init__(self, api_key: str, server_url: str) -> None:
        super().__init__(api_key, server_url)
        self.set_loadout(
            weapon="sword",
            stats={"hp": 3, "speed": 4, "attack": 10, "defense": 3},
            fallback="aggressive",
        )

    async def on_tick(
        self, state: dict, nearby: list, safe_zone: dict
    ) -> dict:
        my_pos: list[int] = state["position"]  # [col, row]

        enemy: dict | None = self.closest_enemy(nearby)
        if enemy is None:
            # No enemies visible - move toward safe zone center.
            return self.move_toward(my_pos, safe_zone["center"])

        dist: int = distance(my_pos, enemy["position"])

        # Within 1 tile (adjacent) - attack.
        if dist < 2:
            return self.attack(enemy["id"])

        # Otherwise charge straight at them.
        return self.move_toward(my_pos, enemy["position"])

    async def on_death(self, death_info: dict) -> None:
        pass  # Berserkers don't mourn.

    async def on_respawn(self, respawn_info: dict) -> None:
        pass  # Back for more.

    async def on_round_end(self, round_info: dict) -> None:
        pass


if __name__ == "__main__":
    key: str = sys.argv[1]
    url: str = sys.argv[2] if len(sys.argv) > 2 else "wss://arena.angel-serv.com/ws/bot"
    bot = BerserkerBot(key, url)
    asyncio.run(bot.run())

File: sdk/nodejs/examples/berserker.js
/**
 * Berserker Bot — charges the closest enemy and attacks relentlessly.
 *
 * Usage:
 *   node berserker.js <API_KEY> [SERVER_URL]
 *
 * Loadout: sword, HP 3 / Speed 4 / Attack 10 / Defense 3
 */

import ArenaBot from '../src/ArenaBot.js';
import { distance } from '../src/helpers.js';

class BerserkerBot extends ArenaBot {
  /**
   * Every tick: find the closest enemy, charge at it, and attack.
   * Positions are [col, row] integers; distances are Chebyshev (tile count).
   * @param {object} state - Our bot's current state.
   * @param {object[]} nearby - Nearby entities.
   * @param {object} safeZone - Safe zone boundaries.
   * @returns {Promise<object>} Action to perform.
   */
  async onTick(state, nearby, safeZone) {
    const enemy = this.closestEnemy(nearby);
    if (!enemy) return this.idle();

    const dist = distance(state.position, enemy.position);

    // Adjacent tile — attack
    if (dist <= 1) {
      return this.attack(enemy.id);
    }

    // Otherwise charge toward the enemy
    return this.moveToward(state.position, enemy.position);
  }

  /** @param {object} deathInfo */
  async onDeath(deathInfo) {
    console.log(
      `[Berserker] Killed by ${deathInfo.killed_by} ` +
      `with ${deathInfo.weapon_used}. Kills this life: ${deathInfo.your_kills_this_life}`
    );
  }

  /** @param {object} roundInfo */
  async onRoundEnd(roundInfo) {
    const stats = roundInfo.your_stats || {};
    console.log(
      `[Berserker] Round ${roundInfo.round_number} — ` +
      `Kills: ${stats.kills ?? 0}, Deaths: ${stats.deaths ?? 0}`
    );
  }
}

// ─── Entry point ──────────────────────────────────────────────────────

const apiKey = process.argv[2];
const serverUrl = process.argv[3];

if (!apiKey) {
  console.error('Usage: node berserker.js <API_KEY> [SERVER_URL]');
  process.exit(1);
}

const bot = new BerserkerBot(apiKey, serverUrl || undefined);
bot.setLoadout('sword', { hp: 3, speed: 4, attack: 10, defense: 3 });
bot.run();
